File: unreal/GahyeonStage/Source/GahyeonRuntimeCore/Private/RealtimeCharacterCoordinator.cpp
#include "Gahyeon/RealtimeCharacterCoordinator.h"

#include <utility>
#include <cmath>
#include <limits>
#include <stdexcept>

namespace Gahyeon {

RealtimeCharacterCoordinator::RealtimeCharacterCoordinator(
    Millis nowMs,
    Millis cognitionTimeoutMs)
    : cognitionTimeoutMs_(cognitionTimeoutMs) {
    if (cognitionTimeoutMs_ <= 0) {
        throw std::invalid_argument("cognition timeout must be positive");
    }
    intents_.Publish(CharacterIntent{
        .Id = "ambient.posture",
        .Layer = IntentLayer::Behavior,
        .Channel = IntentChannel::Posture,
        .GenerationId = std::nullopt,
        .Priority = 1,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = std::nullopt,
        .Value = "ambient_alive",
    });
}

GenerationSyncResult RealtimeCharacterCoordinator::SynchronizeGeneration(Generation generation) {
    const GenerationSyncResult result = intents_.SynchronizeGeneration(generation);
    if (result == GenerationSyncResult::Advanced) {
        thinkingGeneration_.reset();
        thinkingDeadlineMs_.reset();
    }
    return result;
}

Generation RealtimeCharacterCoordinator::VoiceStarted(Millis nowMs) {
    thinkingGeneration_.reset();
    thinkingDeadlineMs_.reset();
    const Generation generation = intents_.BeginGeneration();
    intents_.Publish(CharacterIntent{
        .Id = "conversation.phase",
        .Layer = IntentLayer::Reflex,
        .Channel = IntentChannel::Phase,
        .GenerationId = generation,
        .Priority = 100,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = std::nullopt,
        .Value = "listening",
    });
    intents_.Publish(CharacterIntent{
        .Id = "attention.user." + std::to_string(generation),
        .Layer = IntentLayer::Reflex,
        .Channel = IntentChannel::Attention,
        .GenerationId = generation,
        .Priority = 100,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = 750,
        .Value = "user",
    });
    return generation;
}

PublishResult RealtimeCharacterCoordinator::PartialTranscriptObserved(
    Generation generation,
    double stability,
    Millis nowMs) {
    if (!std::isfinite(stability) || stability < 0.0 || stability > 1.0
        || generation != intents_.CurrentGeneration()) {
        return generation < intents_.CurrentGeneration()
            ? PublishResult::Stale
            : PublishResult::Invalid;
    }
    const ResolvedIntents current = intents_.Resolve(nowMs);
    const CharacterIntent* phase = current.Find(IntentChannel::Phase);
    if (phase == nullptr || phase->Value != "listening") return PublishResult::Invalid;
    return intents_.Publish(CharacterIntent{
        .Id = "attention.user." + std::to_string(generation),
        .Layer = IntentLayer::Reflex,
        .Channel = IntentChannel::Attention,
        .GenerationId = generation,
        .Priority = 100,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = 750,
        .Value = "user",
    });
}

PublishResult RealtimeCharacterCoordinator::VoiceEnded(
    Generation generation,
    Millis nowMs) {
    const PublishResult result = intents_.Publish(CharacterIntent{
        .Id = "conversation.phase",
        .Layer = IntentLayer::Behavior,
        .Channel = IntentChannel::Phase,
        .GenerationId = generation,
        .Priority = 60,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = std::nullopt,
        .Value = "thinking",
    });
    if (result == PublishResult::Accepted) {
        thinkingGeneration_ = generation;
        thinkingDeadlineMs_ = nowMs > std::numeric_limits<Millis>::max() - cognitionTimeoutMs_
            ? std::numeric_limits<Millis>::max()
            : nowMs + cognitionTimeoutMs_;
    }
    return result;
}

bool RealtimeCharacterCoordinator::SpeechStarted(
    Generation generation,
    Millis nowMs,
    std::string utteranceId) {
    const PublishResult phase = intents_.Publish(CharacterIntent{
        .Id = "conversation.phase",
        .Layer = IntentLayer::Cognition,
        .Channel = IntentChannel::Phase,
        .GenerationId = generation,
        .Priority = 70,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = std::nullopt,
        .Value = "speaking",
    });
    if (phase != PublishResult::Accepted) {
        return false;
    }
    thinkingGeneration_.reset();
    thinkingDeadlineMs_.reset();
    return intents_.Publish(CharacterIntent{
        .Id = "conversation.speech",
        .Layer = IntentLayer::Cognition,
        .Channel = IntentChannel::Speech,
        .GenerationId = generation,
        .Priority = 70,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = std::nullopt,
        .Value = std::move(utteranceId),
    }) == PublishResult::Accepted;
}

PublishResult RealtimeCharacterCoordinator::SpeechEnded(
    Generation generation,
    Millis nowMs) {
    const PublishResult speech = intents_.Publish(CharacterIntent{
        .Id = "conversation.speech",
        .Layer = IntentLayer::Behavior,
        .Channel = IntentChannel::Speech,
        .GenerationId = generation,
        .Priority = 10,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = std::nullopt,
        .Value = "",
    });
    if (speech != PublishResult::Accepted) return speech;
    const PublishResult phase = intents_.Publish(CharacterIntent{
        .Id = "conversation.phase",
        .Layer = IntentLayer::Behavior,
        .Channel = IntentChannel::Phase,
        .GenerationId = generation,
        .Priority = 10,
        .CreatedAtMs = nowMs,
        .ExpiresAfterMs = std::nullopt,
        .Value = "idle",
    });
    if (phase == PublishResult::Accepted) {
        thinkingGeneration_.reset();
        thinkingDeadlineMs_.reset();
    }
    return phase;
}

std::optional<Generation> RealtimeCharacterCoordinator::CancelInteraction(
    Generation generation,
    Millis nowMs) {
    if (generation != intents_.CurrentGeneration()) return std::nullopt;
    const Generation next = intents_.BeginGeneration();
    thinkingGeneration_.reset();
    thinkingDeadlineMs_.reset();
    SpeechEnded(next, nowMs);
    return next;
}

std::optional<Generation> RealtimeCharacterCoordinator::Advance(Millis nowMs) {
    if (!thinkingGeneration_.has_value() || !thinkingDeadlineMs_.has_value()
        || nowMs < thinkingDeadlineMs_.value()) {
        return std::nullopt;
    }
    if (thinkingGeneration_.value() != intents_.CurrentGeneration()) {
        thinkingGeneration_.reset();
        thinkingDeadlineMs_.reset();
        return std::nullopt;
    }
    return CancelInteraction(thinkingGeneration_.value(), nowMs);
}

IntentRuntime& RealtimeCharacterCoordinator::Intents() {
    return intents_;
}

const IntentRuntime& RealtimeCharacterCoordinator::Intents() const {
    return intents_;
}

} // namespace Gahyeon

File: desktop/src/runtime/realtime-character-coordinator.ts
import { IntentRuntime } from './intent-runtime'

/**
 * Engine-independent reference coordinator. Unreal C++ should reproduce these
 * transitions while keeping its frame-level animation implementation local.
 */
export class RealtimeCharacterCoordinator {
  readonly intents = new IntentRuntime()
  private thinkingGeneration?: number
  private thinkingDeadlineMs?: number

  constructor(nowMs = 0, private readonly cognitionTimeoutMs = 10_000) {
    if (!Number.isFinite(cognitionTimeoutMs) || cognitionTimeoutMs <= 0)
      throw new Error('cognition timeout must be positive')
    this.intents.publish({
      id: 'ambient.posture',
      layer: 'behavior',
      channel: 'posture',
      priority: 1,
      createdAtMs: nowMs,
      value: 'ambient_alive',
    })
  }

  voiceStarted(nowMs: number) {
    this.clearThinkingDeadline()
    const generation = this.intents.beginGeneration()
    this.intents.publish({
      id: 'conversation.phase',
      layer: 'reflex',
      channel: 'phase',
      generation,
      priority: 100,
      createdAtMs: nowMs,
      value: 'listening',
    })
    this.intents.publish({
      id: `attention.user.${generation}`,
      layer: 'reflex',
      channel: 'attention',
      generation,
      priority: 100,
      createdAtMs: nowMs,
      expiresAfterMs: 750,
      value: 'user',
    })
    return generation
  }

  voiceEnded(generation: number, nowMs: number) {
    const accepted = this.intents.publish({
      id: 'conversation.phase',
      layer: 'behavior',
      channel: 'phase',
      generation,
      priority: 60,
      createdAtMs: nowMs,
      value: 'thinking',
    })
    if (accepted) {
      this.thinkingGeneration = generation
      this.thinkingDeadlineMs = nowMs + this.cognitionTimeoutMs
    }
    return accepted
  }

  speechStarted(generation: number, nowMs: number, utteranceId: string) {
    const phaseAccepted = this.intents.publish({
      id: 'conversation.phase',
      layer: 'cognition',
      channel: 'phase',
      generation,
      priority: 70,
      createdAtMs: nowMs,
      value: 'speaking',
    })
    if (!phaseAccepted) return false
    this.clearThinkingDeadline()
    return this.intents.publish({
      id: 'conversation.speech',
      layer: 'cognition',
      channel: 'speech',
      generation,
      priority: 70,
      createdAtMs: nowMs,
      value: utteranceId,
    })
  }

  speechEnded(generation: number, nowMs: number) {
    const speechAccepted = this.intents.publish({
      id: 'conversation.speech',
      layer: 'behavior',
      channel: 'speech',
      generation,
      priority: 10,
      createdAtMs: nowMs,
      value: '',
    })
    if (!speechAccepted) return false
    const phaseAccepted = this.intents.publish({
      id: 'conversation.phase',
      layer: 'behavior',
      channel: 'phase',
      generation,
      priority: 10,
      createdAtMs: nowMs,
      value: 'idle',
    })
    if (phaseAccepted) this.clearThinkingDeadline()
    return phaseAccepted
  }

  advance(nowMs: number) {
    if (this.thinkingGeneration === undefined || this.thinkingDeadlineMs === undefined
      || nowMs < this.thinkingDeadlineMs) return undefined
    if (this.thinkingGeneration !== this.intents.currentGeneration()) {
      this.clearThinkingDeadline()
      return undefined
    }
    const generation = this.intents.beginGeneration()
    this.clearThinkingDeadline()
    this.speechEnded(generation, nowMs)
    return generation
  }

  private clearThinkingDeadline() {
    this.thinkingGeneration = undefined
    this.thinkingDeadlineMs = undefined
  }
}
